gridstatus/ieso.py

```diff
@@ -390,9 +408,313 @@ def get_zonal_load_forecast(self, date, end=None, verbose=False):
             (pivot_df["Publish Time"] >= date) & (pivot_df["Publish Time"] <= end_date)
         ]
 
-    # TODO add fuel mix. http://reports.ieso.ca/public/GenOutputbyFuelHourly/
     def get_fuel_mix(self, date, end=None, verbose=False):
-        pass
+        """
+        Hourly output and capability for each fuel type (summed over all generators)
+        for a given date or from date to end. Variable generators (solar and wind)
+        have a forecast.
+
+        Args:
+            date (datetime.date | datetime.datetime | str): The date to get the load for
+                Can be a `datetime.date` or `datetime.datetime` object, or a string
+                with the values "today" or "latest". If `end` is None, returns
+                only data for this date.
+            end (datetime.date | datetime.datetime, optional): End date. Defaults None
+                If provided, returns data from `date` to `end` date. The `end` can be a
+                `datetime.date` or `datetime.datetime` object.
+            verbose (bool, optional): Print verbose output. Defaults to False.
+
+        Returns:
+            pd.DataFrame: fuel mix
+        """
+        use_historical = False
+
+        if date != "latest":
+            today = utils._handle_date("today", tz=self.default_timezone)
+            date = utils._handle_date(date, tz=self.default_timezone)
+
+            if date.date() < today.date() - pd.Timedelta(
+                days=REAL_TIME_DAYS_IN_PAST_FUEL_MIX,
+            ):
+                use_historical = True
+            elif date.date() > today.date():
+                raise NotSupported("Fuel mix data is not available for future dates.")
+
+        if use_historical:
+            data = self._retrieve_historical_fuel_mix(date, end, verbose)
+        else:
+            data = (
+                self._retrieve_fuel_mix(date, end, verbose)
+                .groupby(["Fuel Type", "Interval Start", "Interval End"])
+                .sum()
+                .reset_index()
+            )
+
+            pivoted = data.pivot_table(
+                index=["Interval Start", "Interval End"],
+                columns="Fuel Type",
+                values="Output MW",
+            ).reset_index()
+
+            pivoted.columns = [c.title() for c in pivoted.columns]
+            pivoted.index.name = None
+
+            pivoted["Total Output"] = pivoted.sum(numeric_only=True, axis=1)
+
+            data = pivoted.copy()
+
+        return utils.move_cols_to_front(
+            data,
+            [
+                "Interval Start",
+                "Interval End",
+                "Biofuel",
+                "Gas",
+                "Hydro",
+                "Nuclear",
+                "Solar",
+                "Wind",
+                "Total Output",
+            ],
+        )
+
+    def get_generator_output_and_capability(self, date, end=None, verbose=False):
```

Comment:
Example output:

|    | Interval Start      | Interval End        | Generator Name   | Fuel Type   |   Output MW |   Capability MW |   Forecast MW |
|---:|:--------------------|:--------------------|:-----------------|:------------|------------:|----------------:|--------------:|
|  0 | 2023-12-01 01:00:00 | 2023-12-01 02:00:00 | BRUCEA-G1        | NUCLEAR     |         817 |             822 |           nan |
|  1 | 2023-12-01 02:00:00 | 2023-12-01 03:00:00 | BRUCEA-G1        | NUCLEAR     |         817 |             822 |           nan |
|  2 | 2023-12-01 03:00:00 | 2023-12-01 04:00:00 | BRUCEA-G1        | NUCLEAR     |         817 |             822 |           nan |
|  3 | 2023-12-01 04:00:00 | 2023-12-01 05:00:00 | BRUCEA-G1        | NUCLEAR     |         817 |             822 |           nan |
|  4 | 2023-12-01 05:00:00 | 2023-12-01 06:00:00 | BRUCEA-G1        | NUCLEAR     |         817 |             822 |           nan |

Reply:
@kmax12 Do you have an opinion on "capacity" vs "capability"? From the IESO docs:

> * Capability is the maximum potential output of the unit or facility under current conditions, which includes
maximum unit derates and outages for that hour. In this report, Capability is provided for nuclear, hydro, gas,
and biofuel generation (excluding wind and solar).
> * Available Capacity is provided for wind and solar generation. It is the maximum potential output of the
facility, minus derates and outages within the hour. This does not take into account the current availability of
wind and sun (fuel) - which is reflected in the forecast.

The rest of the repo uses "capacity".